evolver/controller/standard/chemostat.py

```diff
@@ -0,0 +1,53 @@
+import time
+from pydantic import Field
+from collections import defaultdict, deque
+from evolver.controller.interface import Controller
+from evolver.hardware.interface import VialConfigBaseModel
+
+
+class Chemostat(Controller):
+    class Config(VialConfigBaseModel):
+        od_sensor: str = Field(description="name of OD sensor to use")
+        pump: str = Field(description="name of pump device to use")
+        stirrer: str = Field(description="name of stirrer device to use")
+        window: int = Field(7, description="number of OD measurements to collect prior to start")
+        start_od: float = Field(0, description="OD at which to start chemostat dilutions")
+        start_time: int = Field(0, description="Time (in hours) after which to start dilutions")
+        flow_rate: float = Field(0, description="Flow rate for dilutions")
+        stir_rate: float = Field(8, description="Stir rate")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # buffer could come from history as well
+        self.od_buffer = defaultdict(lambda: deque(maxlen=self.config.window))
+        # start_time is something we might actually want to be a property of
+        # evolver, in case of interrupt it has a chance of continuing - but can
+        # come from history similarly
+        self.start_time = time.time()
+
+    def control(self, *args, **kwargs):
+        od_val = self.evolver.hardware.get(self.config.od_sensor).get()
+        pump = self.evolver.hardware.get(self.config.pump)
+        stir = self.evolver.hardware.get(self.config.stirrer)
+        elapsed_time = time.time() - self.start_time
+
+        for vial, val in od_val.items():
+            if self.config.vials and vial not in self.config.vials:
+                continue
```

Comment:
this is just to filter the experiment on the configured vials, so no warning is necessary (though might indeed need to error out in the case where a vial is in the config and not provided by OD). 

In the future with a vialed controller and vial view (like https://github.com/ssec-jhu/evolver-ng/pull/30, or the equivalent), this could be implemented with that. But there would always be the possibility to implement a controller against the base hardware, which has multiple vials.

Reply:
>though might indeed need to error out in the case where a vial is in the config and not provided by OD).

👍 

>But there would always be the possibility to implement a controller against the base hardware, which has multiple vials.

Yes, this is a legit scenario that we need to account for. I was concerned with the asymmetry though, what if you add vials but forget to add them to the config, there'd be no warning for this. But maybe there's no way to easily disambiguate between intentionally ignoring vials and unintentionally doing so and that's just user error 🤷 .

Reply:
Maybe there's some rationale here for being more explicit with the ``get`` rather than filtering out vials in the control code, I feel like this is leaky and should happen elsewhere, e.g., in the device code or connection code.

Reply:
I think the vial view concept could help here, and also `get` on vialed devices could prob take a `vial=` argument, but I don't see it can be pushed down at the moment further. We prob have 16 vials - so the OD device also has 16, but if I want to run chemostat on half of them, we just need the controller get the relevant vials for its experiment. For the current API, this is one way to accomplish the filtering.

one important (I think) factor here is running agains all vials without explicitly naming them, that is in essence why it currently asks OD for all vials it has - though it might be better when not configured to ask evolver itself and error if OD isn't configured on one of those.